### PR TITLE
Use StartParameterInternal for GradleInternal.startParameter

### PR DESCRIPTION
**Pull request by wolfs**
This avoids some casts when using `GradleInternal.startParameter`.

### CODE REVIEW

File: subprojects/core/src/main/java/org/gradle/api/internal/StartParameterInternal.java [hunk 1]
```diff
@@ -36,17 +36,17 @@
     private boolean configurationCacheQuiet;
 
     @Override
-    public StartParameter newInstance() {
-        return prepareNewInstance(new StartParameterInternal());
+    public StartParameterInternal newInstance() {
+        return (StartParameterInternal) prepareNewInstance(new StartParameterInternal());
```
Comment (lptr) [inline subprojects/core/src/main/java/org/gradle/api/internal/StartParameterInternal.java:40]: I guess `prepareNewInstance()` could be abstract in `StartParamter` and have an override to return `StartParameterInternal` here.

Also, `prepareNewInstance()` must include copying the internal values like `watchFileSystem`. Maybe that field should move to `StartParameter`? Or we need an override of the method here that also copies internal values? This needs to be taken into account for configuration caching as well. This class hierarchy is a complete mess in any case. :/

Also, maybe rename `prepareNewInstace()` to `makeCopy()` or something, as that's what it does?

File: subprojects/core/src/main/java/org/gradle/api/tasks/GradleBuild.java [hunk 2]
```diff
@@ -186,7 +187,7 @@ public void setBuildName(String buildName) {
     @TaskAction
     void build() {
         // TODO: Allow us to inject plugins into GradleBuild nested builds too.
-        BuildDefinition buildDefinition = BuildDefinition.fromStartParameter(getStartParameter(), getServices().get(PublicBuildPath.class));
+        BuildDefinition buildDefinition = BuildDefinition.fromStartParameter((StartParameterInternal) getStartParameter(), getServices().get(PublicBuildPath.class));
```
Comment (lptr) [inline subprojects/core/src/main/java/org/gradle/api/tasks/GradleBuild.java:190]: This is now a potential `ClassCastException` as a user can call `new StartParameter()` to create a new instance.

Taking a deeper look it seems that we'd already get the exception if an included build was used, but not when included builds are not used. So I guess this is really a new error we are introducing here.

I actually have no good idea how to solve this. :/

One way would be to deprecate the usage of `setStartParameter()`, and do some magic to wrap the value into some sort of delegate for now. I'm not sure it's worth it tbh.

File: subprojects/core/src/main/java/org/gradle/internal/service/scopes/VirtualFileSystemBuildLifecycleListener.java [hunk 3]
```diff
@@ -32,9 +32,10 @@ public VirtualFileSystemBuildLifecycleListener(WatchingAwareVirtualFileSystem vi
         this.virtualFileSystem = virtualFileSystem;
     }
 
+    @SuppressWarnings("deprecation")
```
Comment (lptr) [inline subprojects/core/src/main/java/org/gradle/internal/service/scopes/VirtualFileSystemBuildLifecycleListener.java:35]: Why is this needed now?

File: subprojects/core/src/main/java/org/gradle/api/internal/StartParameterInternal.java (hunk 1 above)
Comment (wolfs) [reply subprojects/core/src/main/java/org/gradle/api/internal/StartParameterInternal.java:40]: `prepareNewInstance` can't be abstract in `StartParameter`, since `StartParameter` isn't an abstract class. Note that `prepareNewInstance` calls `prepareNewBuild`, which in turn copies over the `watchFileSystem` value. So it does copy over the new value, as long as you have a `StartParameterInternal`.

File: subprojects/core/src/main/java/org/gradle/api/tasks/GradleBuild.java (hunk 2 above)
Comment (wolfs) [reply subprojects/core/src/main/java/org/gradle/api/tasks/GradleBuild.java:190]: You will get an exception if you don't use `StartParameterInternal` to run the build. The last moment when that happens is when the provided start parameter is cast to `StartParameterInternal` which happens for example when we create the `buildSrc`-classpath: https://github.com/gradle/gradle/blob/9f988597014407d1569fd81be97cb6651ac785f2/subprojects/core/src/main/java/org/gradle/initialization/buildsrc/BuildSourceBuilder.java#L90

So I don't think casting to `StartParameterInternal` is a breaking change.

File: subprojects/core/src/main/java/org/gradle/internal/service/scopes/VirtualFileSystemBuildLifecycleListener.java (hunk 3 above)
Comment (wolfs) [reply subprojects/core/src/main/java/org/gradle/internal/service/scopes/VirtualFileSystemBuildLifecycleListener.java:35]: To get rid of one more compiler warning for using the deprecated `DEPRECATED_VFS_RETENTION_ENABLED_PROPERTY`. It should have been in there before.

File: subprojects/core/src/main/java/org/gradle/api/tasks/GradleBuild.java (hunk 2 above)
Comment (lptr) [reply subprojects/core/src/main/java/org/gradle/api/tasks/GradleBuild.java:190]: OK. Should we then deprecate `new StartParameter()` and maybe `GradleBuild.setStartParameter()`?

**Comment by wolfs** [comment]
We are not going to do this now, since we are trying to get rid of having a field at all and being able to query for/inject build options directly.
